community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java

```diff
@@ -815,6 +834,37 @@ public void interrupt()
     }
 
     @Override
+    public void markForTermination( Status status, String message )
+    {
+        // NOTE: This is a side-channel method call. You *cannot*
+        //       mutate any of the regular state in the state machine
+        //       from inside this method, it WILL lead to race conditions.
+        //       Imagine this is always called from a separate thread, while
+        //       the main session worker thread is actively working on mutating
+        //       fields on the session.
+        terminationMark.setMark( new Neo4jError( status, message ) );
+
+        // If there is currently a transaction running, terminate it
+        KernelTransaction tx = this.currentTransaction;
+        if(tx != null)
+        {
+            tx.markForTermination();
```

Comment:
Note that #7511 will add a `Status` parameter to the `KernelTransaction.markForTermination` method, which so some extent overlaps with your termination mark feature. Maybe your work here can be simplified in anticipation of #7511.